src/components/shared/thread/ActivityGraphFills.js

```diff
@@ -0,0 +1,653 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+import bisection from 'bisection';
+import clamp from 'clamp';
+
+import './ActivityGraph.css';
+
+import type {
+  IndexIntoSamplesTable,
+  IndexIntoCategoryList,
+  SamplesTable,
+  StackTable,
+} from '../../../types/profile';
+import type {
+  Milliseconds,
+  DevicePixels,
+  CssPixels,
+} from '../../../types/units';
+import type { ActivityGraphProps } from './ActivityGraph';
+
+type SampleContributionToPixel = {|
+  sample: IndexIntoSamplesTable,
+  contribution: number,
+|};
+
+type CategoryFill = {|
+  category: IndexIntoCategoryList,
+  perPixelContribution: Float32Array,
+  fillStyle: string | CanvasPattern,
+|};
+
+export type CategoryDrawStyle = {|
+  +category: number,
+  +gravity: number,
+  +selectedFillStyle: string,
+  +unselectedFillStyle: string,
+  +filteredOutFillStyle: CanvasPattern,
+|};
+
+type SelectedPercentageAtPixelBuffers = {|
+  // The following arrays get recreated when the canvas gets resized.
+  +beforeSelectedPercentageAtPixel: Float32Array,
+  +selectedPercentageAtPixel: Float32Array,
+  +afterSelectedPercentageAtPixel: Float32Array,
+  +filteredOutPercentageAtPixel: Float32Array,
+|};
+
+const BOX_BLUR_RADII = [3, 2, 2];
+const SMOOTHING_RADIUS = 3 + 2 + 2;
+const SMOOTHING_KERNEL: Float32Array = _getSmoothingKernel(
+  SMOOTHING_RADIUS,
+  BOX_BLUR_RADII
+);
+
+/**
+ * A lot of logic and mutation goes into computing the fills in the ThreadActivityGraph.
+ * This class breaks out the fill computations separately from the component logic. This
+ * makes it easier to to have lots of shared mutable state between the various
+ * calculations.
+ *
+ * Note that this class should be recreated for every new computation of the fills.
+ * It computes the category percentages inside a set of Float32Array buffers. These
+ * should be discarded after use. This class has lots of mutation, so there shouldn't
+ * be any assumptions made about the validity of strict equalities.
```

Comment:
I like the idea of factoring out this piece into a bag of information that gets replaced with every draw.

However, I think we can do some more refactoring and reduce the amount of mutation. What I don't like about this solution is the fact that there's both mutating and non-mutating methods, and the class is always used in the sequence 1. create, 2. mutate immediately, 3. never mutate again and use immutably.

I think we can make this more explicit. How about having a free-standing `export function computeFills` which does all the work that `ActivityGraphFills.computeFills` does in your patch, but which then, at the end, creates an instance of `ActivityGraphFills` and returns that. And then only have non-mutating methods on `ActivityGraphFills`, and an accessor for the smoothed cumulative category fills.

As for the actual mutations that happen here, there's actually not that many. You go from a list of category fill styles, to a list of fills with non-cumulative unsmoothed per-pixel-contributions, to a list of fills with non-cumulative *smoothed* per-pixel-contributions, to a list of fills with *cumulative* *smoothed* per-pixel-contributions. So the three mutating operations are 1. computing the contributions, 2. the smoothing and 3. the accumulation. Operations 2 and 3 could just as well be written in non-mutating ways: you might need to recreate the list of fills, and you might need to copy a few typed arrays, but that should be fine. For operation 1, I suppose there's not much you can do. For every sample, you have to look up the category, and then based on the category and the sampleSelectedState, you look up the right percentageBuffer, and then you contribute into that. Yeah, I think this needs to remain a mutating operation.

```
 * This
 * makes it easier to to have lots of shared mutable state between the various
 * calculations.
```

If you make the operations non-mutating as suggested above, then you don't need lots of shared *mutable* state any more, and having all these operations happen from a free-standing `computeFills` function should be fine, I think? Actually, after thinking about operation 1 some more, I'm not sure anymore if that's true... I'd love to hear your thoughts on this.

Reply:
> However, I think we can do some more refactoring and reduce the amount of mutation. What I don't like about this solution is the fact that there's both mutating and non-mutating methods, and the class is always used in the sequence 1. create, 2. mutate immediately, 3. never mutate again and use immutably.

Sounds like a solid plan, I'll try and tangle that out.